cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,860 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+	return HB_MC_SUCCESS;
 }
 
-/*
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor, which must be obtained from hb_mc_init_host.
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_init (uint8_t *fd, eva_id_t eva_id, char *elf, tile_t *tiles, uint32_t num_tiles) {
-	
-	int error = hb_mc_fifo_init(fd); 
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_init() --> hb_mc_init_host(): failed to initialize host.\n");
-		return HB_MC_FAIL;
+int hb_mc_tile_group_allocate_tiles (device_t *device, tile_group_t *tg){
+	int error;
+	if (hb_mc_dimension_get_x(tg->dim) > hb_mc_dimension_get_x(device->mesh->dim)){
+		bsg_pr_err("%s: tile group X dimension (%d) larger than mesh X dimension (%d).\n", __func__, hb_mc_dimension_get_x(tg->dim), hb_mc_dimension_get_x(device->mesh->dim));
+		return HB_MC_INVALID;
 	}
-	
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
-	
-	for (int i = 0; i < num_tiles; i++) { /* initialize tiles */
-		hb_mc_tile_freeze_dep(*fd, tiles[i].x, tiles[i].y);
-		hb_mc_tile_set_group_origin(*fd, tiles[i].x, tiles[i].y, tiles[i].origin_x, tiles[i].origin_y);
+	if (hb_mc_dimension_get_y(tg->dim) > hb_mc_dimension_get_y(device->mesh->dim)){
+		bsg_pr_err("%s: tile group Y dimension (%d) larger than mesh Y dimension (%d).\n", __func__, hb_mc_dimension_get_y(tg->dim), hb_mc_dimension_get_y(device->mesh->dim));
+		return HB_MC_INVALID;
 	}
 
+	// Iterate over the entire mesh as tile (org_y, org_x) being the origin of the new tile group to allcoate 
+	for (hb_mc_idx_t org_y = hb_mc_coordinate_get_y(device->mesh->origin); org_y <= (hb_mc_coordinate_get_y(device->mesh->origin) + hb_mc_dimension_get_y(device->mesh->dim) - hb_mc_dimension_get_y(tg->dim)); org_y++){
+		for (hb_mc_idx_t org_x = hb_mc_coordinate_get_x(device->mesh->origin); org_x <= (hb_mc_coordinate_get_x(device->mesh->origin) + hb_mc_dimension_get_x(device->mesh->dim) - hb_mc_dimension_get_x(tg->dim)); org_x++){
+			int free = 1;
+			hb_mc_idx_t tile_id;
+			
+			// Iterate over a tg->dim.x * tg->dim.y square of tiles starting from the (org_x,org_y) checking to see if all tiles are free so they can be allocated to tile group
+			for (hb_mc_idx_t x = org_x; x < org_x + hb_mc_dimension_get_x(tg->dim); x++){
+				for (hb_mc_idx_t y = org_y; y < org_y + hb_mc_dimension_get_y(tg->dim); y++){
+					tile_id = (y - hb_mc_coordinate_get_y(device->mesh->origin)) * hb_mc_dimension_get_x(device->mesh->dim) + (x - hb_mc_coordinate_get_x(device->mesh->origin));
```

Comment:
Please make a function called something like this:
```c
hb_mc_idx_t get_tile_id(mesh, coordinate)
```
and replace instances of this computation with that function.

Reply:
please make it `static` -- nobody should be calling but you.

Reply:
Max's point is good. To expand on what he's saying (motivate, perhaps) is that people feel that more than three levels of indentation in code is detrimental to reading code. 

FWIW, this is codified in the Linux Kernel Style Guide: 

https://www.kernel.org/doc/html/v4.10/process/coding-style.html#indentation


Something, something, self documenting code... https://en.wikipedia.org/wiki/Self-documenting_code

Reply:
That is more or less my point for some of my other comments. In this comment I’m referring to how ‘tile_id’ is computed.  I guess what it’s doing is finding the index of a tile in some management vector maintained by the mesh? I count five arithmetic ops (which I would say makes this a “complicated” line of code) and this exact line of code seems to occur at multiple lines in this file. That’s why I recommend making it a function. Tell the compiler you want it in-lined if it makes you feel better (but I really wouldn’t worry about that).

Reply:
Fair, I saw the for loops and conflated it with some of your other comments. 

Oh to be in Python and just have an iterator...

Reply:
This all makes perfect sense. To that end, I created a static hb_mc_get_tile_id routine and replaced all instances with this routine in the code. 
I also broke the hb_mc_tile_group_allocate_tiles into three separate routines called hb_mc_tile_group_allocate_tiles, hb_mc_tiles_are_free, and hb_mc_tile_group_initialize_tiles to be more readable. Let me know what you think!